csrc/host_ir/jit.cpp

```diff
@@ -68,185 +78,248 @@ T throwIfError(llvm::Expected<T>&& E) {
-      {int64_ptr_type, int64_type, int64_ptr_type, int64_type, void_ptr_type},
-      false);
-
-  std::string func_name = ir_utils::varName(allocate->buffer()->as<Val>());
-  llvm::Function* func = llvm::Function::Create(
-      func_type,
-      llvm::Function::ExternalLinkage,
-      func_name, // Use the generated unique name
-      mod);
-
-  // Set argument names for better readability in IR
-  func->getArg(0)->setName("sizes");
-  func->getArg(1)->setName("ndim");
-  func->getArg(2)->setName("strides");
-  func->getArg(3)->setName("strides_ndim");
-  func->getArg(4)->setName("out_tensor");
-
-  llvm::BasicBlock* entry = llvm::BasicBlock::Create(context, "entry", func);
-  llvm::IRBuilder<> builder(entry);
-
-  // Get arguments from the function
-  llvm::Value* sizes_arg = func->getArg(0); // const int64_t*
-  llvm::Value* shape_ndim_arg = func->getArg(1); // int64_t (shape ndim)
-  llvm::Value* strides_arg = func->getArg(2); // const int64_t*
-  llvm::Value* strides_ndim_arg = func->getArg(3); // int64_t (strides ndim)
-  llvm::Value* out_tensor_arg = func->getArg(4); // at::Tensor&
-
-  // Bounds checking for ndim
-  auto logical_domain = TensorDomain::noReductions(
-      allocate->buffer()->as<TensorView>()->getLogicalDomain());
-  int64_t ndim = std::ssize(logical_domain);
-  llvm::Value* ndim_val = llvm::ConstantInt::get(int64_type, ndim);
-  // Check shape ndim
-  dimCheck(shape_ndim_arg, ndim_val, builder);
-  // Check strides ndim
-  dimCheck(strides_ndim_arg, shape_ndim_arg, builder);
-
-  // Create constants for type and device from params
-  at::ScalarType data_type = data_type_to_aten(
-      allocate->buffer()->dtype() == DataType::Index
-          ? PrimDataType::Int
-          : allocate->buffer()->dtype());
-  llvm::Value* dtype_constant =
-      llvm::ConstantInt::get(int32_type, static_cast<int32_t>(data_type));
-  llvm::Value* device_index_constant = llvm::ConstantInt::get(
-      int64_type, Communicator::getInstance().deviceId());
-
-  // Get the at::native::empty_strided_cuda function pointer (registered in the
-  // JIT)
-  llvm::Function* at_empty_strided_cuda_func =
-      mod->getFunction(kHostIrJitAtEmptyStridedCudaFuncName);
-  if (at_empty_strided_cuda_func == nullptr) {
-    llvm::FunctionType* at_empty_strided_cuda_func_type =
-        llvm::FunctionType::get(
-            builder.getVoidTy(),
-            {int64_ptr_type,
-             int64_type,
-             int64_ptr_type,
-             int64_type,
-             int32_type,
-             int64_type,
-             void_ptr_type},
-            false);
-    at_empty_strided_cuda_func = llvm::Function::Create(
-        at_empty_strided_cuda_func_type,
-        llvm::Function::ExternalLinkage,
-        kHostIrJitAtEmptyStridedCudaFuncName,
-        mod);
+llvm::Type* getType(JitDataType type, llvm::LLVMContext& ctx) {
+  switch(type) {
+    case JitDataType::void_ptr:
+      return llvm::PointerType::getUnqual(llvm::Type::getInt8Ty(ctx));
```

Comment:
I assume you have access to an IRBuilder. If so, you can ir_builder_->getInt8Ty()->getPointerTo()

Reply:
You still don't like IrBuilder? It seems to be less typing for me.